File: source/python/scripts/generate.py
from __future__ import annotations

import re
import shutil
from pathlib import Path

import grpc_tools
from grpc_tools import protoc


PACKAGE_DIRECTORY = Path(__file__).resolve().parent.parent
REPOSITORY_DIRECTORY = PACKAGE_DIRECTORY.parent.parent
SOURCE_DIRECTORY = PACKAGE_DIRECTORY / "src" / "pole_specification"
PROTO_DIRECTORY = SOURCE_DIRECTORY / "proto"
GENERATED_DIRECTORY = SOURCE_DIRECTORY / "generated"


def main() -> None:
    shutil.rmtree(PROTO_DIRECTORY, ignore_errors=True)
    shutil.rmtree(GENERATED_DIRECTORY, ignore_errors=True)
    PROTO_DIRECTORY.mkdir(parents=True)
    GENERATED_DIRECTORY.mkdir(parents=True)

    proto_files = sorted((REPOSITORY_DIRECTORY / "api").rglob("*.proto"))
    names: set[str] = set()
    for proto_file in proto_files:
        if proto_file.name in names:
            raise RuntimeError(f"duplicate proto basename: {proto_file.name}")
        names.add(proto_file.name)
        shutil.copy2(proto_file, PROTO_DIRECTORY / proto_file.name)

    arguments = [
        "grpc_tools.protoc",
        f"-I{PROTO_DIRECTORY}",
        f"-I{Path(grpc_tools.__file__).resolve().parent / '_proto'}",
        f"--python_out={GENERATED_DIRECTORY}",
        f"--pyi_out={GENERATED_DIRECTORY}",
        f"--grpc_python_out={GENERATED_DIRECTORY}",
        *[str(PROTO_DIRECTORY / name) for name in sorted(names)],
    ]
    if protoc.main(arguments) != 0:
        raise RuntimeError("grpc_tools.protoc generation failed")

    import_pattern = re.compile(r"^import ([A-Za-z0-9_]+_pb2) as ", re.MULTILINE)
    for generated_file in GENERATED_DIRECTORY.glob("*_pb2*.py"):
        content = generated_file.read_text(encoding="utf-8")
        generated_file.write_text(
            import_pattern.sub(r"from . import \1 as ", content),
            encoding="utf-8",
        )

    (GENERATED_DIRECTORY / "__init__.py").write_text("", encoding="utf-8")
    for license_name in ("LICENSE-APACHE", "LICENSE-MIT"):
        shutil.copy2(
            REPOSITORY_DIRECTORY / "source" / "rust" / "pole-specification" / license_name,
            PACKAGE_DIRECTORY / license_name,
        )


if __name__ == "__main__":
    main()

File: source/node/scripts/generate.mjs
import { cp, mkdir, readdir, rm, writeFile } from "node:fs/promises";
import { basename, dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { spawnSync } from "node:child_process";

const packageDirectory = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const repositoryDirectory = resolve(packageDirectory, "../..");
const apiDirectory = join(repositoryDirectory, "api");
const protoDirectory = join(packageDirectory, "proto");
const generatedDirectory = join(packageDirectory, "src", "generated");

async function collectProtoFiles(directory) {
  const entries = await readdir(directory, { withFileTypes: true });
  const files = [];
  for (const entry of entries) {
    const path = join(directory, entry.name);
    if (entry.isDirectory()) {
      files.push(...await collectProtoFiles(path));
    } else if (entry.isFile() && entry.name.endsWith(".proto")) {
      files.push(path);
    }
  }
  return files.sort();
}

await rm(protoDirectory, { recursive: true, force: true });
await rm(generatedDirectory, { recursive: true, force: true });
await mkdir(protoDirectory, { recursive: true });
await mkdir(generatedDirectory, { recursive: true });

const sourceFiles = await collectProtoFiles(apiDirectory);
const seenNames = new Set();
for (const sourceFile of sourceFiles) {
  const name = basename(sourceFile);
  if (seenNames.has(name)) {
    throw new Error(`duplicate proto basename: ${name}`);
  }
  seenNames.add(name);
  await cp(sourceFile, join(protoDirectory, name));
}

for (const licenseName of ["LICENSE-APACHE", "LICENSE-MIT"]) {
  await cp(
    join(repositoryDirectory, "source", "rust", "pole-specification", licenseName),
    join(packageDirectory, licenseName),
  );
}

const generator = join(packageDirectory, "node_modules", ".bin", "proto-loader-gen-types");
const result = spawnSync(generator, [
  "--includeComments",
  "--includeDirs",
  protoDirectory,
  "--grpcLib",
  "@grpc/grpc-js",
  "--outDir",
  generatedDirectory,
  ...[...seenNames].sort().map((name) => join(protoDirectory, name)),
], { stdio: "inherit" });

if (result.status !== 0) {
  process.exit(result.status ?? 1);
}

const manifest = [...seenNames].sort();
await writeFile(
  join(packageDirectory, "src", "proto-files.json"),
  JSON.stringify(manifest, null, 2) + "\n",
);
